grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java

```diff
@@ -559,10 +537,11 @@ public RelationshipAtom inferTypes() {
      */
     private RelationshipAtom inferRoles(Answer sub){
         //return if all roles known and non-meta
-        Set<Role> explicitRoleTypes = getExplicitRoles();
-        boolean allRolesMeta = explicitRoleTypes.stream().filter(role -> Schema.MetaSchema.isMetaLabel(role.getLabel())).count() == getRelationPlayers().size();
-        boolean metaRoleRecomputationViable = allRolesMeta && !sub.isEmpty();
-        if (explicitRoleTypes.size() == getRelationPlayers().size() && !metaRoleRecomputationViable) return this;
+        List<Role> explicitRoles = getExplicitRoles().collect(Collectors.toList());
+        Map<Var, Type> varTypeMap = getParentQuery().getVarTypeMap();
+        boolean allRolesMeta = explicitRoles.stream().filter(role -> Schema.MetaSchema.isMetaLabel(role.getLabel())).count() == getRelationPlayers().size();
```

Comment:
There's a method `allMatch`:
```
boolean allRolesMeta = explicitRoles.stream().allMatch(role -> 
    Schema.MetaSchema.isMetaLabel(role.getLabel())
);
```
which I think does the same thing (if `getRelationPlayers()` is always the same size as `explicitRoles`)